lib/database/repositories/QcFlagRepository.js

```diff
@@ -156,107 +88,70 @@ class QcFlagRepository extends Repository {
-                GROUP_CONCAT( DISTINCT qcfv.flag_id ) AS verifiedFlagsList,
-                GROUP_CONCAT( DISTINCT qcf.id ) AS flagsList
-
-            FROM quality_control_flags AS qcf
-            INNER JOIN quality_control_flag_types AS qcft
-                ON qcft.id = qcf.flag_type_id
-            LEFT JOIN quality_control_flag_verifications AS qcfv
-                ON qcfv.flag_id = qcf.id
-            INNER JOIN quality_control_flag_effective_periods AS qcfep
-                ON qcf.id = qcfep.flag_id
-            INNER JOIN data_pass_quality_control_flag AS dpqcf
-                ON dpqcf.quality_control_flag_id = qcf.id
-            INNER JOIN (${GAQ_PERIODS_VIEW}) AS gaq_periods
-                ON gaq_periods.data_pass_id = dpqcf.data_pass_id
-            INNER JOIN global_aggregated_quality_detectors AS gaqd
-                ON gaqd.data_pass_id = gaq_periods.data_pass_id
-                    AND gaqd.run_number = gaq_periods.run_number
-                    AND gaqd.detector_id = qcf.detector_id
-                    AND gaq_periods.run_number = qcf.run_number
-                        AND (qcfep.\`from\` IS NULL OR UNIX_TIMESTAMP(qcfep.\`from\`) <= gaq_periods.\`from\`)
-                        AND (qcfep.\`to\`   IS NULL OR gaq_periods.\`to\` <= UNIX_TIMESTAMP(qcfep.\`to\`))
-
-            GROUP BY
-                gaq_periods.data_pass_id,
-                gaq_periods.run_number,
-                gaq_periods.\`from\`,
-                gaq_periods.\`to\`
-            `;
-
         const query = `
             SELECT
-                effectivePeriods.runNumber,
-                effectivePeriods.dataPassId,
-                effectivePeriods.bad,
-                SUM(effectivePeriods.mcReproducible) > 0 AS mcReproducible,
-                GROUP_CONCAT(effectivePeriods.verifiedFlagsList) AS verifiedFlagsList,
-                GROUP_CONCAT(effectivePeriods.flagsList) AS flagsList,
+                gaq_periods.runNumber,
+                gaq_periods.dataPassId,
+                gaq_periods.bad,
+                gaq_periods.badWhenMcReproducibleAsNotBad,
+                SUM(gaq_periods.mcReproducible) > 0 AS mcReproducible,
+                GROUP_CONCAT(gaq_periods.verifiedFlagsList) AS verifiedFlagsList,
+                GROUP_CONCAT(gaq_periods.flagsList) AS flagsList,
 
                 IF(
                     (
-                        COALESCE(run.time_trg_end,   run.time_o2_end  ) IS NULL
-                     OR COALESCE(run.time_trg_start, run.time_o2_start) IS NULL
+                        run.time_end   IS NULL
+                     OR run.time_start IS NULL
                     ),
                     IF(
                         SUM(
-                            COALESCE(effectivePeriods.\`to\`  , 0)
-                          + COALESCE(effectivePeriods.\`from\`, 0)
+                            COALESCE(gaq_periods.\`to\`  , 0)
+                          + COALESCE(gaq_periods.\`from\`, 0)
```

Comment:
No, you need to have some grouping function there, in other words, this `if` covers case, when all `gaq eff. periods ` have nullish timestamps (so we have only one period)

Reply:
I am not sure to understand. For me `0` is very different to null. If I am correct here what you want to have is 1 if both are null, else null, so it should be explicit?

Reply:
yes, but if and only if there is no mixture of defined and undefined flags boundaries.
Example 1: If you have run with undefined end, period `(10:00, null)` and some QC flag covering period `(10:00, 11:00)` and another one `(11:00, null)`, then you cannot calculate coverage

Example 2:  If you have run with both boundaries undefined , period (null, null) and some QC flag covering period `(null, 11:00)` and another one `(11:00, null)`, then you cannot calculate coverage

You can calculate coverage (GAQ and ordinary QC as well) if and only if,
    * case 1 ) all timestamps (of a run and flags) are null, btw then you can have only one `gaq_effective_period` `(null, null)` (or qc_flag_eff._period respectively), because all flags are run-based
    * case 2 )  all timestamps (of a run and flags) are defined

case 3 ) mixed state

If we have `case 3` then we would have e.g. gaq_periods like `(null, 11:00)` and `(11:00, null)`. And yes regardless which of the gaq_periods goes to condition you proposed, results should be always correct, I am not 100% sure, because  it's mixing of aggregation and non-aggregation. I would say that the way we aggregate should be based on whole batch of data (all gaq_periods), but not on one sample, and I think what you try to do is exactly that.

Reply:
applied